v2/pubsub-to-elasticsearch/src/test/java/com/google/cloud/teleport/v2/templates/PubSubToElasticsearchTest.java

```diff
@@ -0,0 +1,266 @@
+                testAttributeMap2));
+
+    nonFilterableTestMessages =
+        ImmutableList.of(
+            makePubsubMessage(
+                "{\"location\":\"MH\", \"name\":\"Lara\", \"age\":\"35\", \"color\":\"black\", \"coffee\":\"mochachino\"}",
+                null,
+                null),
+            new PubsubMessage("".getBytes(), testAttributeMap1));
+
+    badTestMessages =
+        ImmutableList.of(
+            makePubsubMessage("This is a bad record", null, null),
+            makePubsubMessage("with unknown attribute", "dummy", "value"));
+
+    allTestMessages =
+        ImmutableList.<PubsubMessage>builder()
+            .addAll(goodTestMessages)
+            .addAll(badTestMessages)
+            .addAll(nonFilterableTestMessages)
+            .build();
+  }
+
+  /** Tests the {@link PubSubToElasticsearch} pipeline end-to-end with no UDF supplied. */
+  @Test
+  public void testPubSubToElasticsearchNoUdfE2E() {
+
+    FailsafeElementCoder<String, String> failsafeElementCoder =
+        PubSubToElasticsearch.FAILSAFE_ELEMENT_CODER;
+
+    CoderRegistry coderRegistry = pipeline.getCoderRegistry();
+
+    coderRegistry.registerCoderForType(
+        failsafeElementCoder.getEncodedTypeDescriptor(), failsafeElementCoder);
+
+    coderRegistry.registerCoderForType(
+        PubSubToElasticsearch.CODER.getEncodedTypeDescriptor(), PubSubToElasticsearch.CODER);
+
+    PubSubToElasticsearch.PubSubToElasticsearchOptions options =
+        TestPipeline.testingPipelineOptions()
+            .as(PubSubToElasticsearch.PubSubToElasticsearchOptions.class);
+
+    options.setDeadletterTable("test:dataset.table");
+    options.setJavascriptTextTransformFunctionName(null);
+    options.setJavascriptTextTransformGcsPath(null);
+
+    PCollectionTuple pc =
+        pipeline
+            .apply(Create.of(goodTestMessages.get(0)))
+            .apply(
+                PubSubToElasticsearch.PubSubMessageToJsonDocument.newBuilder()
+                    .setJavascriptTextTransformFunctionName(
+                        options.getJavascriptTextTransformFunctionName())
+                    .setJavascriptTextTransformGcsPath(options.getJavascriptTextTransformGcsPath())
+                    .build());
+
+    PAssert.that(pc.get(PubSubToElasticsearch.TRANSFORM_OUT))
+        .satisfies(
+            collection -> {
+              FailsafeElement<PubsubMessage, String> element = collection.iterator().next();
+              assertThat(
+                  element.getOriginalPayload().getPayload(),
+                  is(equalTo(goodTestMessages.get(0).getPayload())));
+              return null;
+            });
+
+    // Execute pipeline
+    pipeline.run(options);
+  }
+
+  /** Tests the {@link PubSubToElasticsearch} pipeline end-to-end with a UDF. */
+  @Test
+  public void testPubSubToElasticsearchUdfE2E() {
+    FailsafeElementCoder<String, String> failsafeElementCoder =
+        PubSubToElasticsearch.FAILSAFE_ELEMENT_CODER;
```

Comment:
Why not directly use FAILSAFE_ELEMENT_CODER instead of assigning it to failsafeElementCoder?